src/main/java/com/bushidowallet/core/bitcoin/tx/Transaction.java

```diff
@@ -1,356 +1,395 @@
+        return ByteUtil.reverseBytes(Hash.hash(uncheckedSerialize()));
+    }
+
+    public Transaction from(UTXODescriptor utxoDescriptor) throws Exception {
+        List<UTXODescriptor> utxos = new ArrayList<UTXODescriptor>();
+        utxos.add(utxoDescriptor);
+        return from(utxos);
+    }
+
+    public Transaction from(List<UTXODescriptor> utxos) throws Exception {
+
+        return from(utxos, null, 0);
+    }
+
+    public Transaction from(List<UTXODescriptor> utxos, List<ECKey> publicKeys, int treshhold) throws Exception {
+
+        for (int i = 0; i < utxos.size(); i++) {
+            if (utxoIncluded(utxos.get(i)) == true) {
+                throw new Exception("utxos list contains element that has been already converted to transaction input.");
+            }
+            if (publicKeys != null && publicKeys.size() > 0 && treshhold > 0) {
+                fromMultisigUTXO(utxos.get(i), publicKeys, treshhold);
+            } else {
+                fromNonP2SH(utxos.get(i));
+            }
+        }
+        return this;
+    }
+
+    public Transaction change(Address address) throws Exception {
+        this.changeScript = Script.fromAddress(address);
+        updateChangeOutput();
+        return this;
+    }
+
+    private void updateChangeOutput() throws Exception {
+        if (changeScript == null) {
+            return;
+        }
+        clearSignatures();
+        if (changeIndex >= 0) {
+            this.outputs.remove(changeIndex);
+        }
+        long available = this.getUnspentValue();
+        long fee = getFee();
+        long changeAmount = available - fee;
+        if (changeAmount > 0) {
+            changeIndex = this.outputs.size();
+            addOutput(new Output(changeScript, changeAmount));
+        } else {
+            changeIndex = -1;
+        }
+    }
+
+    private void clearSignatures() throws Exception {
+        for (Input input : inputs) {
+            input.clearSignatures();
+        }
+    }
+
+    public Transaction setFee(long fee) throws Exception {
+        this.fee = fee;
+        updateChangeOutput();
+        return this;
+    }
+
+    public long getFee() throws Exception {
+        if (changeScript == null) {
+            return getUnspentValue();
+        }
+        return fee == -1 ? estimateFee() : fee;
+    }
+
+	public long estimateFee() throws Exception {
+		long size = estimateSize();
+		return new BigDecimal(size).divide(new BigDecimal(1024), 10, BigDecimal.ROUND_HALF_DOWN)
+				.multiply(new BigDecimal(FEE_PER_KB)).longValue();
+
+		/*
+		 * long available = getUnspentValue(); long size = estimateSize();
+		 * BigDecimal feeBase = new BigDecimal(size).divide(new
+		 * BigDecimal(FEE_PER_KB)); double fee =
+		 * Math.ceil(feeBase.doubleValue()); if (available > fee) { size +=
+		 * CHANGE_OUTPUT_MAX_SIZE; } return (long) Math.ceil(new
+		 * BigDecimal(size).divide(new BigDecimal(1000)).doubleValue() *
+		 * FEE_PER_KB);
+		 */

```

Comment:
I have ported this from https://github.com/bitpay/bitcore-lib/blob/master/lib/transaction/transaction.js line 915. I do see You are proposing some simplification here. Are You open to discuss this @liguanliang2014?